_SPIKE_.md

```diff
@@ -0,0 +1,297 @@
+and information on different approaches I have researched and abandoned along
+the way.
+
+## Overview
+
+In this feature branch, I am presenting a proof of concept implementation
+demonstrating how to describe navigational properties for inclusion of related
+models.
+
+The solution consists of three parts:
+
+1. A convention for describing navigational properties (links) via an interface,
+   including support in `CrudRepository` and `DefaultCrudRepository` classes.
+
+2. Improvements in repository-json-schema (`getJsonSchema` and
+   `modelToJsonSchema`):
+
+   - 2.1: A new option `includeRelations` controlling whether navigational
+     properties are included in the generated schema.
+   - 2.2: Support for cyclic references. For example: `CategoryWithLinks` has a
+     property `products` containing an array of `ProductWithLinks`.
+     `ProductWithLinks` has a property `category` containing
+     `CategoryWithLinks`.
+   - 2.3: A new helper function `modelToJsonSchemaRef` that emits JSON Schema
+     reference and includes definitions of all references models.
+
+3. Improvements in code generating controller OpenAPI spec allowing app
+   developers to use `modelToJsonSchemaRef` to describe response schema.
+
+Additionally, the todo-list example has been updated to show the proposed
+solution in practice.
+
+## Discussion
+
+We have the following requirements on a solution for describing navigational
+properties:
+
+1. A (compile-time) type describing navigational properties of a model. For
+   example, when a `Category` model has many instances of a `Product` model,
+   then we want to define property `products` containing an array of `Product`
+   instances _with product navigational properties included_.
+
+2. Ability to define a compile-time type where all navigational properties are
+   defined as optional. This is needed by Repository implementation.
+
+3. Ability to generate two OpenAPI/JSON Schemas for each model:
+
+   1. Own properties only
+   2. Both own properties and navigational properties
+
+4. SHOULD HAVE: To support JavaScript developers and declarative support, the
+   new types should be optional. At runtime, we should leverage the dynamic
+   nature of JavaScript objects and add navigational properties to an instance
+   of the original mo/contrdel. Specifically, we should not require another
+   model class to represent model with links.
+
+My proposed solution meets all requirements above. Additionally, it consists of
+several smaller building blocks that can be used beyond the scope of
+navigational properties too.
+
+## Solution details
+
+### Interface describing model links
+
+To describe navigation properties for TypeScript compiler, application
+developers will define a new interface for each model.
+
+BelongsTo relation:
+
+```ts
+@model()
+class Product extends Entity {
+  @belongsTo(() => Category)
+  categoryId: number;
+}
+
+interface ProductLinks {
```

Comment:
In the doc, you use `ProductWithLinks`.

Reply:
I assume that by docs, you mean the user story _2.2: Support for cyclic references._? The example there is just an acceptance criteria to drive that particular fix and allow us to implement  necessary changes before we start working on navigational properties for relations.

I'll clarify the docs.